File: web/src/lib/calendar.ts
/*
 * Single owner of "which calendar day is this?".
 *
 * Two stored shapes:
 *  - instant — created_at / updated_at / resolved_at / status_changed_at /
 *    reopened_at / assignee_changed_at. UTC timestamps. Calendar day is the
 *    day in the given IANA zone (the viewer's local zone in the UI).
 *  - date — duedate. YYYY-MM-DD stored as written. Never Date.parse it
 *    (UTC midnight would print the previous day in the Americas).
 *
 * Must stay aligned with internal/calendar (same cases in both test files).
 */

export type DateKind = 'instant' | 'date'

export interface CalendarZone {
  id: string
}

export function localZone(): CalendarZone {
  try {
    return { id: Intl.DateTimeFormat().resolvedOptions().timeZone || 'UTC' }
  } catch {
    return { id: 'UTC' }
  }
}

export function utcZone(): CalendarZone {
  return { id: 'UTC' }
}

export function zoneNamed(id: string): CalendarZone {
  return { id }
}

const YMD = /^(\d{4}-\d{2}-\d{2})/

export interface CalendarDecision {
  raw: string
  kind: DateKind
  zone: string
  calendar_day: string
  ok: boolean
}

/** Report how calendarDay classified raw. */
export function explain(
  raw: string | null | undefined,
  kind: DateKind,
  zone: CalendarZone = localZone(),
): CalendarDecision {
  const day = calendarDay(raw, kind, zone)
  return {
    raw: raw ?? '',
    kind,
    zone: zone.id,
    calendar_day: day ?? '',
    ok: day != null,
  }
}

/** YYYY-MM-DD for a stored value, or null when unreadable. */
export function calendarDay(
  raw: string | null | undefined,
  kind: DateKind,
  zone: CalendarZone = localZone(),
): string | null {
  if (raw == null) return null
  const s = raw.trim()
  if (!s) return null
  const ymd = YMD.exec(s)?.[1] ?? null
  if (kind === 'date') return ymd
  if (ymd && s.length === 10) return ymd
  const d = new Date(s)
  if (Number.isNaN(d.getTime())) return ymd
  return formatYmd(d, zone.id)
}

/**
 * Inclusive YYYY-MM-DD compare after calendarDay.
 * Empty raw fails when any bound is set; with no bounds it still passes.
 */
export function inRange(
  raw: string | null | undefined,
  kind: DateKind,
  from: string | null | undefined,
  to: string | null | undefined,
  zone: CalendarZone = localZone(),
): boolean {
  if (!raw) return !from && !to
  const day = calendarDay(raw, kind, zone)
  if (!day) return false
  if (from && day < from) return false
  if (to && day > to) return false
  return true
}

/** ISO date of this week's Monday in zone. */
export function startOfWeekMonday(now: Date = new Date(), zone: CalendarZone = localZone()): string {
  const weekday = new Intl.DateTimeFormat('en-US', { timeZone: zone.id, weekday: 'short' }).format(now)
  const offset = { Mon: 0, Tue: 1, Wed: 2, Thu: 3, Fri: 4, Sat: 5, Sun: 6 }[weekday]
  if (offset == null) return calendarDay(now.toISOString(), 'instant', zone) ?? ''
  const today = calendarDay(now.toISOString(), 'instant', zone)
  if (!today) return ''
  const [y, m, d] = today.split('-').map(Number)
  const mon = new Date(Date.UTC(y, m - 1, d - offset))
  const yy = mon.getUTCFullYear()
  const mm = String(mon.getUTCMonth() + 1).padStart(2, '0')
  const dd = String(mon.getUTCDate()).padStart(2, '0')
  return `${yy}-${mm}-${dd}`
}

/**
 * Absolute display. Date-only strings stay on their calendar day (no UTC
 * midnight). Instants use the given zone.
 */
export function formatAbs(
  raw: string | null | undefined,
  kind: DateKind,
  zone: CalendarZone = localZone(),
  locale = 'en-US',
): string {
  if (!raw) return ''
  const s = raw.trim()
  if (kind === 'date' || (s.length === 10 && YMD.test(s))) {
    const day = calendarDay(s, 'date', zone)
    if (!day) return raw
    const [y, m, d] = day.split('-').map(Number)
    return new Date(y, m - 1, d).toLocaleDateString(locale, {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
    })
  }
  const d = new Date(s)
  if (Number.isNaN(d.getTime())) return raw
  return d.toLocaleString(locale, {
    timeZone: zone.id,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
  })
}

function formatYmd(d: Date, timeZone: string): string {
  const parts = new Intl.DateTimeFormat('en-CA', {
    timeZone,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).formatToParts(d)
  const year = parts.find((p) => p.type === 'year')?.value
  const month = parts.find((p) => p.type === 'month')?.value
  const day = parts.find((p) => p.type === 'day')?.value
  if (!year || !month || !day) return ''
  return `${year}-${month}-${day}`
}

File: internal/calendar/calendar.go
// Package calendar is the single owner of "which calendar day is this?".
//
// Two stored shapes exist:
//
//   - Instant — created_at, updated_at, resolved_at, status_changed_at,
//     reopened_at, assignee_changed_at. UTC timestamps. The calendar day is
//     the day in the given zone (local for the UI / Match).
//   - Date — duedate. A YYYY-MM-DD calendar date stored as written. Never
//     parse it as UTC midnight.
//
// Go and the web (web/src/lib/calendar.ts) implement the same rules.
// Explain reports the decision so a mismatch is observable.
package calendar

import (
	"strings"
	"time"
)

// Kind selects how a stored value is read.
type Kind int

const (
	// Instant is a UTC timestamp (RFC3339 / ISO-8601).
	Instant Kind = iota
	// Date is a calendar day stored as YYYY-MM-DD.
	Date
)

func (k Kind) String() string {
	if k == Date {
		return "date"
	}
	return "instant"
}

// Zone is the calendar used for Instant → day. Date values ignore it.
type Zone struct {
	loc *time.Location
}

// Local is the process local zone (the viewer's zone in the UI / CLI).
func Local() Zone { return Zone{time.Local} }

// UTC is the UTC zone.
func UTC() Zone { return Zone{time.UTC} }

// In wraps a location. nil falls back to Local.
func In(loc *time.Location) Zone {
	if loc == nil {
		return Local()
	}
	return Zone{loc}
}

// Named loads an IANA zone (tests pin Asia/Seoul so CI UTC cannot hide a miss).
func Named(name string) (Zone, error) {
	loc, err := time.LoadLocation(name)
	if err != nil {
		return Zone{}, err
	}
	return Zone{loc}, nil
}

func (z Zone) locOrLocal() *time.Location {
	if z.loc == nil {
		return time.Local
	}
	return z.loc
}

// Name is the IANA / Local identifier Explain reports.
func (z Zone) Name() string {
	return z.locOrLocal().String()
}

// Decision is one Day conversion, for tests and debugging.
type Decision struct {
	Raw         string `json:"raw"`
	Kind        string `json:"kind"`
	Zone        string `json:"zone"`
	CalendarDay string `json:"calendar_day"`
	OK          bool   `json:"ok"`
}

// Explain reports how Day classified raw.
func Explain(raw string, kind Kind, z Zone) Decision {
	d := Decision{Raw: raw, Kind: kind.String(), Zone: z.Name()}
	day, ok := Day(raw, kind, z)
	d.CalendarDay = day
	d.OK = ok
	return d
}

// Day returns YYYY-MM-DD for raw, or false when it cannot be read.
func Day(raw string, kind Kind, z Zone) (string, bool) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "", false
	}
	if kind == Date {
		return dateOnly(raw)
	}
	if ymd, ok := dateOnly(raw); ok && len(raw) == 10 {
		// Date-only string passed as Instant: do not parse as UTC midnight.
		return ymd, true
	}
	t, ok := parseInstant(raw)
	if !ok {
		return dateOnly(raw)
	}
	return t.In(z.locOrLocal()).Format("2006-01-02"), true
}

// InRange is an inclusive YYYY-MM-DD compare after Day.
// An empty raw fails when any bound is set; with no bounds it still passes.
func InRange(raw string, kind Kind, from, to string, z Zone) bool {
	if strings.TrimSpace(raw) == "" {
		return strings.TrimSpace(from) == "" && strings.TrimSpace(to) == ""
	}
	day, ok := Day(raw, kind, z)
	if !ok {
		return false
	}
	if from != "" && day < from {
		return false
	}
	if to != "" && day > to {
		return false
	}
	return true
}

// StartOfWeekMonday is the ISO date of Monday 00:00 in z for now.
func StartOfWeekMonday(now time.Time, z Zone) string {
	t := now.In(z.locOrLocal())
	offset := int(t.Weekday()+6) % 7 // Monday = 0
	mon := time.Date(t.Year(), t.Month(), t.Day()-offset, 0, 0, 0, 0, z.locOrLocal())
	return mon.Format("2006-01-02")
}

// FormatDay is YYYY-MM-DD of t in z. compileDate uses this so JQL
// startOfDay() / -7d land on the same calendar the matcher uses.
func FormatDay(t time.Time, z Zone) string {
	return t.In(z.locOrLocal()).Format("2006-01-02")
}

func dateOnly(raw string) (string, bool) {
	if len(raw) < 10 {
		return "", false
	}
	ymd := raw[:10]
	if _, err := time.Parse("2006-01-02", ymd); err != nil {
		return "", false
	}
	return ymd, true
}

func parseInstant(raw string) (time.Time, bool) {
	layouts := []string{
		time.RFC3339Nano,
		time.RFC3339,
		"2006-01-02T15:04:05.000Z07:00",
		"2006-01-02T15:04:05Z07:00",
		"2006-01-02 15:04:05Z07:00",
		"2006-01-02 15:04:05",
	}
	for _, layout := range layouts {
		if t, err := time.Parse(layout, raw); err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}
